micropython.js

```diff
@@ -147,10 +151,21 @@ class MicroPythonBoard {
 
   async run(code, data_consumer) {
     data_consumer = data_consumer || function() {}
-    await this.enter_raw_repl()
-    const output = await this.exec_raw(code || '#', data_consumer)
-    await this.exit_raw_repl()
-    return Promise.resolve(output)
+    return new Promise(async (resolve, reject) => {
+      if (this.reject_run) {
+        this.reject_run('re run')
+        this.reject_run = null
```

Comment:
This line is not necessary here, but if you want to keep setting the rejected promise to `null`, maybe you can have a method for all three occurrences.

```js
function ensureRejected(message) {
  if (this.reject_run) {
    this.reject_run(new Error(message));
    this.reject_run = null;
  }
}
```

Any reason you prefer rejecting with a `string` instead of an `Error`?

Reply:
No, it should definitely be an error. I just don't find the helper function really helpful here. I'm also not sure if this is going to be the *final solution* :tm: so I'd prefer keeping things a bit more unrolled.